server/resource_server/repository/repository_test.go

```diff
@@ -288,3 +306,309 @@ func TestProfileClient(t *testing.T) {
+	).WithArgs(
+		"true",
+		sqlmock.AnyArg(),
+		userId,
+		"email_verified",
+	).WillReturnResult(
+		sqlmock.NewResult(0, 1),
+	)
+
+	mock.ExpectCommit()
+
+	err = repository.SaveProofOfEmailVerification(userId, verificationCode, emailProof)
+
+	assert.Nil(t, err)
+
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("There were unfulfilled expectations: %s", err)
+	}
+}
+
+func TestSaveEmailVerificationData_RowWithUserIdDoesNotExist(t *testing.T) {
+	SetUp(t)
+	defer mockDB.Close()
+
+	mock.ExpectBegin()
+
+	mock.ExpectQuery(
+		regexp.QuoteMeta(`SELECT * FROM "email_verification_data" WHERE "email_verification_data"."user_id" = $1 ORDER BY "email_verification_data"."id" LIMIT 1`),
+	).WithArgs(
+		userId,
+	).WillReturnRows(
+		sqlmock.NewRows(
+			[]string{"id", "user_id", "verification_code", "expires_at"},
+		),
+	)
+
+	mock.ExpectQuery(
+		regexp.QuoteMeta(
+			`INSERT INTO "email_verification_data" ("user_id","verification_code","expires_at") VALUES ($1,$2,$3) RETURNING "id"`,
+		),
+	).WithArgs(
+		userId, verificationCode, timestamp,
+	).WillReturnRows(
+		sqlmock.NewRows(
+			[]string{"id", "user_id", "verification_code", "expires_at"},
+		).AddRow(
+			1, userId, verificationCode, timestamp,
+		),
+	)
+
+	mock.ExpectCommit()
+
+	err = repository.SaveEmailVerificationData(
+		models.EmailVerificationData{
+			UserId:           userId,
+			VerificationCode: verificationCode,
+			ExpiresAt:        timestamp,
+		},
+	)
+
+	if err != nil {
+		t.Fatalf("Error while saving email verification data: %v", err)
+	}
+	if err = mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("There were unfulfilled expectations: %s", err)
+	}
+}
+
+func TestSaveEmailVerificationData_RowWithUserIdExists(t *testing.T) {
+	SetUp(t)
+	defer mockDB.Close()
+
+	mock.ExpectBegin()
+
+	mock.ExpectQuery(
+		regexp.QuoteMeta(
+			`SELECT * FROM "email_verification_data" WHERE "email_verification_data"."user_id" = $1 ORDER BY "email_verification_data"."id" LIMIT 1`,
+		),
+	).WithArgs(
+		userId,
+	).WillReturnRows(
+		sqlmock.NewRows(
+			[]string{"id", "user_id", "verification_code", "expires_at"},
+		).AddRow(
+			id, userId, verificationCode, timestamp,
+		),
+	)
+
+	mock.ExpectExec(
+		regexp.QuoteMeta(
+			`UPDATE "email_verification_data" SET "expires_at"=$1,"user_id"=$2,"verification_code"=$3 WHERE "email_verification_data"."user_id" = $4 AND "id" = $5`,
+		),
+	).WithArgs(
+		timestamp, userId, verificationCode, userId, id,
+	).WillReturnResult(sqlmock.NewResult(2, 1))
+
+	mock.ExpectCommit()
+
+	err = repository.SaveEmailVerificationData(
+		models.EmailVerificationData{
+			UserId:           userId,
+			VerificationCode: verificationCode,
+			ExpiresAt:        timestamp,
+		},
+	)
+
+	if err != nil {
+		t.Fatalf("Error while saving email verification data: %v", err)
+	}
+	if err = mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("There were unfulfilled expectations: %s", err)
+	}
+}
+
+func TestGetEmailVerificationData(t *testing.T) {
```

Comment:
I have a hard time giving this advice because you've already written better testing than anyone else on the team. However, technically, we are missing this test's reverse: `TestGetEmailVerificationData_Failure`

Reply:
Added it!